File: en/codes/csharp/chapter_hashing/built_in_hash.cs
/**
* File: built_in_hash.cs
* Created Time: 2023-06-26
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_hashing;

public class built_in_hash {
    [Test]
    public void Test() {
        int num = 3;
        int hashNum = num.GetHashCode();
        Console.WriteLine("The hash value of integer " + num + " is" + hashNum);

        bool bol = true;
        int hashBol = bol.GetHashCode();
        Console.WriteLine("The hash value of boolean " + bol + " is" + hashBol);

        double dec = 3.14159;
        int hashDec = dec.GetHashCode();
        Console.WriteLine("The hash value of decimal " + dec + " is" + hashDec);

        string str = "Hello algorithm";
        int hashStr = str.GetHashCode();
        Console.WriteLine("The hash value of string " + str + " is" + hashStr);

        object[] arr = [12836, "Ha"];
        int hashTup = arr.GetHashCode();
        Console.WriteLine("Array [" + string.Join(", ", arr) + "] hash value is" + hashTup);

        ListNode obj = new(0);
        int hashObj = obj.GetHashCode();
        Console.WriteLine("The hash value of node object " + obj + " is" + hashObj);
    }
}

File: en/codes/cpp/chapter_hashing/built_in_hash.cpp
/**
 * File: built_in_hash.cpp
 * Created Time: 2023-06-21
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    int num = 3;
    size_t hashNum = hash<int>()(num);
    cout << "Integer " << num << " has hash value " << hashNum << "\n";

    bool bol = true;
    size_t hashBol = hash<bool>()(bol);
    cout << "Boolean " << bol << " has hash value " << hashBol << "\n";

    double dec = 3.14159;
    size_t hashDec = hash<double>()(dec);
    cout << "Decimal " << dec << " has hash value " << hashDec << "\n";

    string str = "Hello Algo";
    size_t hashStr = hash<string>()(str);
    cout << "String " << str << " has hash value " << hashStr << "\n";

    // In C++, built-in std::hash() only provides hash calculation for basic data types
    // Hash calculation for arrays and objects needs to be implemented manually
}
